### DISCUSSION TITLE
[pnpm monorepo(w/ turborepo)] Unable to find any GraphQL type definitions for the following pointers:

### DISCUSSION THREAD

**Discussion by nagisa-yamato** [graphql-hive/graphql-eslint#1835] [Q&A]
Hi,

I am trying to set up graphql-eslint in a pnpm monorepo using turborepo.

- Each workspace under `/apps` uses separate GraphQL endpoints
- graphql-codegen is installed in the root of monorepo ([as recommended in docs](https://the-guild.dev/graphql/codegen/docs/getting-started/development-workflow#monorepo-and-yarn-workspaces)) 

However, I get the error below when I try to run eslint (CLI and VSCode extension).

```jsh
~/projects/yqninzlnb.github
❯ pnpm eslint .
Pages directory cannot be found at /home/projects/yqninzlnb.github/pages or /home/projects/yqninzlnb.github/src/pages. If using a custom path, please configure with the `no-html-link-for-pages` rule in your eslint config file.

/home/projects/yqninzlnb.github/apps/docs/app/page.tsx
  0:0  error  Parsing error: [graphql-eslint] 
      Unable to find any GraphQL type definitions for the following pointers:
        
          - ./app/**/*.graphql

/home/projects/yqninzlnb.github/apps/web/app/page.tsx
  0:0  error  Parsing error: [graphql-eslint] 
      Unable to find any GraphQL type definitions for the following pointers:
        
          - ./app/**/*.graphql

/home/projects/yqninzlnb.github/codegen.ts
  19:8  error  Prefer named exports  import/no-default-export

✖ 3 problems (3 errors, 0 warnings)


~/projects/yqninzlnb.github 5s
❯ 
```

Is there a solution to fix this error?

- [reproduction repository](https://github.com/nagisa-yamato/turborepo-graphql-eslint)
- [above running on StackBlitz](https://stackblitz.com/github/nagisa-yamato/turborepo-graphql-eslint?file=README.md)

**Comment by nagisa-yamato** [comment]
When I tried switching to `graphql.config.ts`, I got graphql-eslint to work but graphql-codegen failed with the error `[ProjectNotFoundError: Project 'default' not found]`.

Is using `graphql.config.ts` the proper way to fix this issue?
Is there a problem with graphql-codegen or graphql-config?

```jsh
~/projects/vbwiloljr.github 5s
❯ pnpm lint --filter web

> turborepo-graphql-eslint@ lint /home/projects/vbwiloljr.github
> turbo run lint "--filter" "web"

Issues occurred when constructing package graph. Turbo will function, but some features may not be available Error: Unable to generate pnpm-lock.yaml with lockfileVersion: 6.0. Supported lockfile versions are 5.3,5.4
• Packages in scope: web
• Running lint in 1 packages
• Remote caching disabled
web:lint:  
web:lint:  > web@1.0.0 lint /home/projects/vbwiloljr.github/apps/web
web:lint:  > eslint . --ignore-path .gitignore
web:lint:  
web:lint:  
web:lint:  /home/projects/vbwiloljr.github/apps/web/app/page.tsx
web:lint:    11:11  error  Cannot query field "foo" on type "Film"  @graphql-eslint/fields-on-correct-type
web:lint:  
web:lint:  ✖ 1 problem (1 error, 0 warnings)
web:lint:  
web:lint:   ELIFECYCLE  Command failed with exit code 1.
 ERROR  run failed: running web#lint failed: command finished with error

 Tasks:    0 successful, 1 total
Cached:    0 cached, 1 total
  Time:    4.1s

 ELIFECYCLE  Command failed with exit code 1.

~/projects/vbwiloljr.github 4s
❯ pnpm generate

> turborepo-graphql-eslint@ generate /home/projects/vbwiloljr.github
> graphql-codegen

[ProjectNotFoundError: Project 'default' not found]
 ELIFECYCLE  Command failed with exit code 1.

~/projects/vbwiloljr.github
❯ 
```

- [reproduction repository](https://github.com/nagisa-yamato/turborepo-graphql-eslint/tree/test/graphql-config)
  - `graphql.config.ts`
    ```ts
    import type { CodegenConfig } from "@graphql-codegen/cli";
    import type { IGraphQLConfig } from "graphql-config";
    
    const webCodegenConfig: CodegenConfig = {
      ignoreNoDocuments: true,
      generates: {
        "packages/codegen/web/": {
          preset: "client",
        },
      },
    };
    
    const docsCodegenConfig: CodegenConfig = {
      ignoreNoDocuments: true,
      generates: {
        "packages/codegen/docs/": {
          preset: "client",
        },
      },
    };
    
    const config: IGraphQLConfig = {
      projects: {
        web: {
          documents: "apps/web/**/!(*.d).{ts,tsx}",
          schema: "https://swapi-graphql.netlify.app/.netlify/functions/index",
          extensions: {
            codegen: webCodegenConfig,
          },
        },
        docs: {
          documents: "apps/docs/**/!(*.d).{ts,tsx}",
          schema: "https://rickandmortyapi.com/graphql",
          extensions: {
            codegen: docsCodegenConfig,
          },
        },
      },
    };
    
    export default config;
    ```
- [above running on StackBlitz](https://stackblitz.com/github/nagisa-yamato/turborepo-graphql-eslint/tree/test/graphql-config?file=README.md)

**Reply by dimaMachina** [reply]
`next lint` doesn't works with overrides https://github.com/B2o5T/graphql-eslint/issues/1095#issuecomment-1160907823

**Reply by nagisa-yamato** [reply]
Hi, thank you for the reply.

> next lint doesn't works with overrides https://github.com/B2o5T/graphql-eslint/issues/1095#issuecomment-1160907823

Yes, I've noticed this and changed the npm scripts to use `eslint .` instead, making graphql-eslint to catch errors, when I use `graphql.config.ts`.

To clarify, here are the results I am getting right now.

| config              | eslint result                                                                                           | codegen result                                      | reproduction repository                                                            | stackblitz                                                                                                   |
| ------------------- | ------------------------------------------------------------------------------------------------------- | --------------------------------------------------- | ---------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------ |
| `codegen.ts`        | Parsing error: [graphql-eslint] Unable to find any GraphQL type definitions for the following pointers: | success                                             | https://github.com/nagisa-yamato/turborepo-graphql-eslint                          | https://stackblitz.com/github/nagisa-yamato/turborepo-graphql-eslint?file=README.md                          |
| `graphql.config.ts` | success                                                                                                 | [ProjectNotFoundError: Project 'default' not found] | https://github.com/nagisa-yamato/turborepo-graphql-eslint/tree/test/graphql-config | https://stackblitz.com/github/nagisa-yamato/turborepo-graphql-eslint/tree/test/graphql-config?file=README.md |

Is using `graphql.config.ts` the proper way to fix this issue?
Is there a problem with graphql-codegen or graphql-config?

**Reply by dimaMachina** [reply]
please share a branch when you use `codegen.ts` and `graphql.config.ts`

**Reply by nagisa-yamato** [reply]
Can you access these two branches?

- branch using `codegen.ts` (main)
  - https://github.com/nagisa-yamato/turborepo-graphql-eslint
- branch using `graphql.config.ts` (test/graphql-config)
  - https://github.com/nagisa-yamato/turborepo-graphql-eslint/tree/test/graphql-config

Or do you mean a branch that uses BOTH `codegen.ts` and `graphql.config.ts`?

**Reply by dimaMachina** [reply]
I am in your main branch and I receive the following error

```sh
/Users/dmytro/Desktop/repros/turborepo-graphql-eslint/apps/docs/app/page.tsx
  0:0  error  Parsing error: [graphql-eslint] 
      Unable to find any GraphQL type definitions for the following pointers:
        
          - ./app/**/*.graphql
```

the problem is because you specify

```json5
      parserOptions: {
        operations: "./app/**/*.graphql",
```

and graphql-tools can't find any definition in the following files, you must use `./app/**/*.tsx` instead and you will receive

```sh
/Users/dmytro/Desktop/repros/turborepo-graphql-eslint/apps/docs/app/page.tsx
  10:9  error  Field `id` defined multiple times  @graphql-eslint/no-duplicate-fields

✖ 1 problem (1 error, 0 warnings)

```

**Reply by dimaMachina** [reply]
also, you must specify https://the-guild.dev/graphql/eslint/docs/getting-started/parser-options#skipgraphqlconfig when specifying documents/schema inside parserOptions, otherwise parserOptions are ignored and top-level graphql.config.ts is loaded

**Reply by nagisa-yamato** [reply]
> you must use ./app/**/*.tsx

So I changed `parserOptions.operations` of the eslint config, and I'm still getting the same error.

- https://github.com/nagisa-yamato/turborepo-graphql-eslint/tree/fix/operations
- https://stackblitz.com/github/nagisa-yamato/turborepo-graphql-eslint/tree/fix/operations?file=README.md

```sh
~/projects/jvmizgzmk.github
❯ pnpm eslint .
Pages directory cannot be found at /home/projects/jvmizgzmk.github/pages or /home/projects/jvmizgzmk.github/src/pages. If using a custom path, please configure with the `no-html-link-for-pages` rule in your eslint config file.

/home/projects/jvmizgzmk.github/apps/docs/app/page.tsx
  0:0  error  Parsing error: [graphql-eslint] 
      Unable to find any GraphQL type definitions for the following pointers:
        
          - ./app/**/*.tsx

/home/projects/jvmizgzmk.github/apps/web/app/page.tsx
  0:0  error  Parsing error: [graphql-eslint] 
      Unable to find any GraphQL type definitions for the following pointers:
        
          - ./app/**/*.tsx

/home/projects/jvmizgzmk.github/codegen.ts
  19:8  error  Prefer named exports  import/no-default-export

✖ 3 problems (3 errors, 0 warnings)


~/projects/jvmizgzmk.github 4s
❯ 
```

I've applied changes to

- `./apps/docs/.eslintrc.js`
- `./apps/web/.eslintrc.js`

Am I applying changes to the wrong eslint config file?

**Reply by nagisa-yamato** [reply]
- specifying `skipGraphQLConfig` did not change the result
- the branch I tested does not have a `graphql.config.ts` file

**Reply by dimaMachina** [reply]
@nagisa-yamato I checked your fix/operations branch and everything fine on my side

```sh
 dmytro@MacBook-Pro-3  ~/Desktop/repros/turborepo-graphql-eslint   fix/operations  pnpm lint --force

> turborepo-graphql-eslint@ lint /Users/dmytro/Desktop/repros/turborepo-graphql-eslint
> turbo run lint "--force"

• Packages in scope: codegen, docs, eslint-config-custom, tsconfig, ui, web
• Running lint in 6 packages
• Remote caching disabled
ui:lint: cache bypass, force executing 3b93c1b9fc42d7c3
docs:lint: cache bypass, force executing 66b50cc7329abcdd
web:lint: cache bypass, force executing 34f14427d7062778
ui:lint: 
ui:lint: > ui@0.0.0 lint /Users/dmytro/Desktop/repros/turborepo-graphql-eslint/packages/ui
ui:lint: > eslint .
ui:lint: 
docs:lint: 
docs:lint: > docs@1.0.0 lint /Users/dmytro/Desktop/repros/turborepo-graphql-eslint/apps/docs
docs:lint: > eslint . --ignore-path .gitignore
docs:lint: 
web:lint: 
web:lint: > web@1.0.0 lint /Users/dmytro/Desktop/repros/turborepo-graphql-eslint/apps/web
web:lint: > eslint . --ignore-path .gitignore
web:lint: 

 Tasks:    3 successful, 3 total
Cached:    0 cached, 3 total
  Time:    7.7s
```

try to run turbo with `--force` flag

**Reply by dimaMachina** [reply]
same without turbo

```sh
 dmytro@MacBook-Pro-3  ~/Desktop/repros/turborepo-graphql-eslint   fix/operations  cd apps/web
 dmytro@MacBook-Pro-3  ~/Desktop/repros/turborepo-graphql-eslint/apps/web   fix/operations  pnpm lint        

> web@1.0.0 lint /Users/dmytro/Desktop/repros/turborepo-graphql-eslint/apps/web
> eslint . --ignore-path .gitignore

 dmytro@MacBook-Pro-3  ~/Desktop/repros/turborepo-graphql-eslint/apps/web   fix/operations  

```

**Reply by nagisa-yamato** [reply]
I've noticed that

- running `turbo run lint --force` from project root
- running `eslint .` in each workspace (ex. `/apps/web`)

does not throw an error (as in the logs you provided above).

However, I am still having the same errors when running _**`eslint .` at the project root**_.

The reason for this is 

1. There is a separate project with a similar structure using [`lint-staged`](https://github.com/okonet/lint-staged), and I have to [run eslint directly (NOT use `turbo run ...`)](https://github.com/vercel/turbo/issues/4445#issuecomment-1507724210) to integrate with it.
2. The error is still present on the VSCode ESLint extension.

![Screenshot 2023-08-25 at 12 52 11](https://github.com/B2o5T/graphql-eslint/assets/110368475/cdbc0a63-d6d2-43f5-8bf4-a99c2dcd1024)

---

The current workaround I found is using both `graphql.config.ts` and `codegen.ts`, and specifying `codegen.ts` as the config when running graphql-codegen.

Ideally, it would be better to unify the config to `graphql.config.ts` because they are redundant settings, but I can't get codegen to run with only `graphql.config.ts` ([as mentioned before](https://github.com/B2o5T/graphql-eslint/discussions/1835#discussioncomment-6811510), please check the reproduction repo or StackBlitz).

Will there be a better workaround?
